### ISSUE TITLE
[naming-convention] Destructuring assignment matches the "property" selector instead of "variable"

### ISSUE THREAD

**Issue by octogonz** [typescript-eslint/typescript-eslint#2250]
Regarding the `property` selector, the [naming-convention docs](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/naming-convention.md) say:

> - `property` - matches any object, class, or object type property. Does not match properties that have direct function expression or arrow function expression values.
>  - Allowed `modifiers`: `private`, `protected`, `public`, `static`, `readonly`, `abstract`.
>  - Allowed `types`: `boolean`, `string`, `number`, `function`, `array`.

However it incorrectly matches `CONSTANT1` in the expression below:

```ts
// Why does CONSTANT1 match the "property" selector? 
// It is a "variable" in this context.
const { CONSTANT1, CONSTANT2 } = someLibrary.constants;
```

**Additional Info**

This was originally discussed in https://github.com/typescript-eslint/typescript-eslint/issues/2244#issuecomment-648603915 but we decided that it is a separate issue.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `3.3.0` |
| `@typescript-eslint/parser`        | `3.3.0` |
| `TypeScript`                       | `3.5.3` |
| `ESLint`                           | `7.2.0` |
| `node`                             | `12.17.0` |
| `npm`                              | `6.14.4` |

**Comment by seiyab** [comment]
I wonder if I can configure they match variable or property.

**Comment by bradzacher** [comment]
Destructuring is not a property declaration, it's syntactic sugar for declaring a variable.
When this is fixed, it will not match property any more, and it will not be configurable.

Copying comment from other thread:

---

> Why does CONSTANT1 match the "property" selector? 

I believe it'll actually try to match both the variable and property selectors.
This is a bug. If you wouldn't mind filing a new issue for this.

The bug occurs because `Property` is reused for both object literals and object destructuring, and the code doesn't ignore properties in a destructuring context.

This should pull the identifiers from the destructuring and analyse them as a variable
https://github.com/typescript-eslint/typescript-eslint/blob/5fdd21a1726fb6928098c4152aec55a30df960d4/packages/eslint-plugin/src/rules/naming-convention.ts#L433-L445
	
And this should pull the identifiers as a property
https://github.com/typescript-eslint/typescript-eslint/blob/5fdd21a1726fb6928098c4152aec55a30df960d4/packages/eslint-plugin/src/rules/naming-convention.ts#L518-L523

**Comment by seiyab** [comment]
Thank you for your detailed explanation.
> Destructuring is not a property declaration

That's true. I understand matching property doesn't make sense.

So what I need is ignoring destructuring assignment.
I might open another issue if I really need it.

**Comment by Dar0n** [comment]
Hello,

I'll let myself to jump in and add a bit of complexity to the original example. 
What if I use a destructuring with renaming? As in:
`const { bad_name: goodName } = someLibrary.object`
In this case bad_name should not be linted at all, as it is the name used in the object outside of my control. So, it should actually lint only the goodName part.

However, I am not familiar with the internals of linting process. Is there a way to correctly resolve this?
Thank you!

**Comment by bodinsamuel** [comment]
Hello,

+1 on @Dar0n comment. This is big blocker on legacy codebase. 
Especially when values comes from database you can not have camelcase at all, 
we always use destructuring+renaming which do not work anymore.

Would love an option to enable this behavior ☺️

**Comment by paztis** [comment]
Hi

It was possible previously to support it in old camelCase rule (v2.x), with the parameter ignoreDestructuring": true
Here you've remove this rule and don't propose equivalent solution...


It force me to fully disable your new rule...

**Comment by bradzacher** [comment]
Fixed in #2566
